### DISCUSSION TITLE
Aheadworks Product quiestion module integration

### DISCUSSION THREAD

**Discussion by pk-pankaj** [breezefront/community#41] [1. Help]
Hi There,

I am trying to integrate the "Aheadwork Pquesion" module to make breeze compatible.

there are used some dependency in a file like below:
define([
    'jquery',
    'underscore',
    'Magento_Ui/js/form/form',
    'mage/translate'
], function ($, _, Component, $t) {
    'use strict';
 
Here, how we can use "underscore" and "Magento_Ui/js/form/form" dependencies while integrating due to this form is not being generated?

Thanks

**Comment by vovayatsyuk** [comment]
You need to rewrite this file without Magento_Ui/js/form/form dependency.

A possible replacement for this component is the following view component inherited from built-in 'form' component:

```js
$.view('pQuestion', 'form', {
    component: 'AW_Pquestion/js/[..view/form..]',
    ...
});
```

**Reply by pk-pankaj** [reply]
Thanks for your reply! it helped me.

Are there any components for the below: 
'Magento_Ui/js/form/element/abstract',
'Magento_Ui/js/form/element/textarea',
'Magento_Ui/js/form/element/single-checkbox',
'Magento_Ui/js/form/provider'

Thanks.

**Reply by vovayatsyuk** [reply]
No, these are not available. I recommend rewriting the js component to use a phtml template (contact form for example) instead of ko-driven html's. This is a much better way to write a fast CLS-free magento frontend.

**Reply by pk-pankaj** [reply]
We will look to go with your suggestion.

Thanks @vovayatsyuk

**Reply by pk-pankaj** [reply]
We have another issues:
`.observe(
                    [
                        'isLoading',
                        'ratingValue',
                        'actionState',
                        'ratingValueAdditionalClass'
                    ]
                );`

this seems does not work giving errors when we are using like below in the same file

`this.ratingValueAdditionalClass("");`

I am attaching the full file code, can you please review if is it possible to integrate this with the breeze?

`define([
    'underscore',
    'jquery',
    'uiComponent',
    'Aheadworks_Pquestion/js/action/rating/get-metadata',
    'Aheadworks_Pquestion/js/action/rating/update'
], function (_, $, Component, getRatingMetadataAction, updateRatingAction) {
    'use strict';

    return Component.extend({
        defaults: {
            template: 'Aheadworks_Pquestion/rating/component',
            isLoading: false,
            entityId: '',
            entityType: '',
            ratingValue: 0,
            ratingValueAdditionalClass: '',
            actionState: 0,
            isCustomerLoggedIn: false,
            canGuestChangeRating: false,
            imports: {
                entityId: '${ $.dataProvider }:data.entity_id',
                entityType: '${ $.dataProvider }:data.entity_type',
                isCustomerLoggedIn: '${ $.configProvider }:data.is_customer_logged_in',
                canGuestChangeRating: '${ $.configProvider }:data.can_guest_change_rating'
            }
        },
        votedClass: 'voted',

        /**
         * @inheritdoc
         */
        initialize: function () {
            this._super()
                ._updateRatingValueClass()
            ;

            this.isLoading(true);
            getRatingMetadataAction(this.entityType, this.entityId)
                .then(function (response) {
                    this._processRatingMetadata(response);
                    this.isLoading(false);
                }.bind(this));

            return this;
        },

        /**
         * @inheritdoc
         */
        initObservable: function () {
            this._super()
                .observe(
                    [
                        'isLoading',
                        'ratingValue',
                        'actionState',
                        'ratingValueAdditionalClass'
                    ]
                );

            var self = this;

            this.ratingValue.subscribe(function () {
                self._updateRatingValueClass();
            });

            return this;
        },

        /**
         * Check if change rating action is available for current visitor
         *
         * @returns {boolean}
         */
        isChangeRatingActionAvailable: function () {
            return this.isCustomerLoggedIn || this.canGuestChangeRating;
        },

        /**
         * On like click
         */
        onLikeClick: function () {
            this._changeRating(1);
        },

        /**
         * On dislike click
         */
        onDislikeClick: function () {
            this._changeRating(-1);
        },

        /**
         * Get additional CSS class for like button
         *
         * @returns {string}
         */
        getLikeButtonAdditionalClass: function () {
            var additionalClass = '';

            if (this.actionState() > 0) {
                additionalClass = this.votedClass;
            }
            return additionalClass;
        },

        /**
         * Get additional CSS class for dislike button
         *
         * @returns {string}
         */
        getDislikeButtonAdditionalClass: function () {
            var additionalClass = '';

            if (this.actionState() < 0) {
                additionalClass = this.votedClass;
            }
            return additionalClass;
        },

        /**
         * Change rating by action type
         *
         * @param actionType
         * @private
         */
        _changeRating: function (actionType) {
            this.isLoading(true);
            updateRatingAction(this.entityType, this.entityId, actionType)
                .then(function (response) {
                    this._processRatingMetadata(response);
                    this.isLoading(false);
                }.bind(this));
        },

        /**
         * Update CSS additional class for rating value
         *
         * @private
         */
        _updateRatingValueClass: function () {
            this.ratingValueAdditionalClass('');
            if (this.ratingValue() > 0) {
                this.ratingValueAdditionalClass('positive-value')
            } else if (this.ratingValue() < 0) {
                this.ratingValueAdditionalClass('negative-value')
            }
        },

        /**
         * Process rating metadata within the response
         *
         * @param response
         * @private
         */
        _processRatingMetadata: function (response) {
            if (response) {
                if (!_.isUndefined(response.rating_value)) {
                    this.ratingValue(response.rating_value);
                }
                if (!_.isUndefined(response.action_state)) {
                    this.actionState(response.action_state);
                }
            }
        }
    });
});
`

**Reply by vovayatsyuk** [reply]
`this.observe` is a shorthand to create `ko.observables`. Here is how I implemented it in one of the integrations: https://github.com/breezefront/module-breeze-amasty-xsearch/blob/master/view/frontend/web/js/wrapper.js#L55-L71

p.s. You will also have to implement `imports` functionality, because it doesn't exist in breeze.

**Reply by pk-pankaj** [reply]
Hi @vovayatsyuk 

Thanks for your reply.

how we can create and implement `imports` functionality, is there any reference?

**Reply by vovayatsyuk** [reply]
No, I do not have any examples of `imports` implementation in breeze. 

In order to implement it you need to know what are dataProvider and configProvider are referenced to. You can find this info in the component initialization json string (in HTML source code).

Unfortunately, it's not possible to provide more information without seeing the module sources.

**Reply by pk-pankaj** [reply]
Thanks @vovayatsyuk 
we will try.
